optaplanner-core/src/main/java/org/optaplanner/core/impl/score/inliner/ScoreInliner.java

```diff
@@ -16,18 +16,96 @@
 
 package org.optaplanner.core.impl.score.inliner;
 
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
 import org.optaplanner.core.api.score.Score;
+import org.optaplanner.core.api.score.constraint.ConstraintMatch;
+import org.optaplanner.core.api.score.constraint.ConstraintMatchTotal;
+import org.optaplanner.core.api.score.constraint.Indictment;
+import org.optaplanner.core.api.score.stream.Constraint;
+import org.optaplanner.core.impl.score.constraint.DefaultConstraintMatchTotal;
+import org.optaplanner.core.impl.score.constraint.DefaultIndictment;
 
 public abstract class ScoreInliner<Score_ extends Score<Score_>> {
 
+    private final Map<String, Score_> constraintIdToWeightMap;
     protected final boolean constraintMatchEnabled;
+    private final Score_ zeroScore;
+    private final Map<String, DefaultConstraintMatchTotal<Score_>> constraintMatchTotalMap;
+    private final Map<Object, DefaultIndictment<Score_>> indictmentMap;
 
-    protected ScoreInliner(boolean constraintMatchEnabled) {
+    protected ScoreInliner(Map<Constraint, Score_> constraintToWeightMap, boolean constraintMatchEnabled,
+            Score_ zeroScore) {
+        this.constraintIdToWeightMap = Objects.requireNonNull(constraintToWeightMap).entrySet().stream()
+                .collect(Collectors.toMap(e -> e.getKey().getConstraintId(), Map.Entry::getValue));
         this.constraintMatchEnabled = constraintMatchEnabled;
+        this.zeroScore = zeroScore;
+        this.constraintMatchTotalMap = constraintMatchEnabled ? new LinkedHashMap<>() : null;
+        this.indictmentMap = constraintMatchEnabled ? new LinkedHashMap<>() : null;
     }
 
     public abstract Score_ extractScore(int initScore);
 
-    public abstract WeightedScoreImpacter buildWeightedScoreImpacter(Score_ constraintWeight);
+    /**
+     * Create a new instance of {@link WeightedScoreImpacter} for a particular constraint.
+     * 
+     * @param constraint never null
+     * @return never null
+     */
+    public abstract WeightedScoreImpacter buildWeightedScoreImpacter(Constraint constraint);
```

Comment:
ScoreImpacter now is constraint-specific and it is necessary to construct a new one each time the constraint weight changes.
Since yesterday (commit 0a086a2fdb11e732cc66bb427fe073934fbbec28), we now have tests to make sure that this actually happens.